### PR TITLE
DS-3286 Add automatic matching algorithm for calculations [revdep skip]

### PR DESCRIPTION
**Pull request by jrwishart**
Added automatic matching algorithm as specified in DS-3256.

### CODE REVIEW

File: R/utilities.R [hunk 1]
```diff
@@ -1121,13 +1127,13 @@ createDimNames <- function(names.required, index, n.dim)
...
 #' dimension p times.
 #' @param dims A list containing an integer vector of the dimensions of the two
 #' inputs. E.g. the above example would have list(c(n, p), c(n, 1))
-#' @return A list with two elements. \code{NULL} if the input isn't to be reshaped.
+#' @return A list with two elements. \code{NULL} if the input isn't to be recycled.
 #' Otherwise contains a sublist with two elements called, \itemize{
-#' \item dims.required : The new dimensions of the reshaped element
+#' \item dims.required : The new dimensions of the recycled element
 #' \item dim.to.rep : Which dimension to do the reshaping. Takes the value 1 if
 #' the row dimension it to be repeated and 2 if the column element to be repeated.
```
Comment (mwmclean) [inline R/utilities.R:1138]: Typos

**Review by mwmclean — CHANGES_REQUESTED** [review]
1. Getting an error when I try to use fuzzy matching [here](https://staff.displayr.com/Dashboard?project_id=-848743#page=5256d8ff-5a14-400a-9b25-cc837615f815)

2. Getting an error trying to add a p x p matrix and a p-vector with colnames matching the matrix [here](https://staff.displayr.com/Dashboard?project_id=-848743#page=5fcf9951-dbb7-43bf-aaae-285e4f86d442)

**Comment by jrwishart** [comment]
> 1. Getting an error when I try to use fuzzy matching [here](https://staff.displayr.com/Dashboard?project_id=-848743#page=5256d8ff-5a14-400a-9b25-cc837615f815)
> 2. Getting an error trying to add a p x p matrix and a p-vector with colnames matching the matrix [here](https://staff.displayr.com/Dashboard?project_id=-848743#page=5fcf9951-dbb7-43bf-aaae-285e4f86d442)

For 1. this is by design, the error message could be better but the option should be a named vector. i.e. change 
```
Sum(m1.3, m2.3, match.elements = c(match.opts[3], match.opts[3]))
```
to 
```
Sum(m1.3, m2.3, match.elements = c(match.rows = match.opts[3], match.columns = match.opts[3]))
```

For 2. it is an edge case where one of the inputs has a missing na (i.e. the last element of the named vector `vec` has `NA` for its last name). I could fix this to hide any elements with `NA` as their name when hiding unmatched is requested and giving an error if there is more than one name with `NA` when showing matches is requested.

File: R/sum.R [hunk 2]
```diff
@@ -16,18 +16,36 @@
 #'   structures to be removed from the column dimension of the input. Any column elements
 #'   with the labels specified here will not be included in the resulting \code{Sum}
 #'   calculation.
-#' @param match.rows The names of the row elements of the inputs are inspected and compared.
-#'   If matches are found then the input will permute the rows of the elements so that the
-#'   elements match in the row dimension. There are five options available for row matching.
+#' @param match.elements Either a single character string with three possible options or named character vector with two elements. The possible single character options are: \itemize{
+#'   \item "No": Ignores names and requires either inputs which the same dimensions
```
[the remark this answers is not in the document]
Comment (jrwishart) [reply R/sum.R:20]: Ok yep, I'll update to make that clearer

**Comment by jrwishart** [comment]
For 1. I could change it to either
1. accept a unnamed argument of length 2 and assume the first is for rows and 2nd for columns or
2. I could force them to provide the two names. I was going to use this option but the checking function I wrote earlier didn't catch the unnamed option properly and manifested as an error here.

**Comment by mwmclean** [comment]
Oops, PEBKAC. 1. sounds fine to me, regarding supporting an unnamed vector for `match.elements`.

If a user has multiple rows/columns with NA names and requests to show unmatched, shouldn't it just include them unmatched in the output without erroring?

**Review by mwmclean — APPROVED** [review]

**Pull request merged**